pkg/ipamd/ipamd.go

```diff
@@ -743,26 +743,30 @@ func (c *IPAMContext) setupENI(eni string, eniMetadata awsutils.ENIMetadata, tru
 	if err != nil && err.Error() != datastore.DuplicatedENIError {
 		return errors.Wrapf(err, "failed to add ENI %s to data store", eni)
 	}
+	// Store the primary IP of the ENI
+	c.primaryIP[eni] = eniMetadata.PrimaryIPv4Address()
 
 	// For secondary ENIs, set up the network
 	if eni != c.awsClient.GetPrimaryENI() {
-		eniPrimaryIP := eniMetadata.PrimaryIPv4Address()
-		err = c.networkClient.SetupENINetwork(eniPrimaryIP, eniMetadata.MAC, eniMetadata.DeviceNumber, eniMetadata.SubnetIPv4CIDR)
+		err = c.networkClient.SetupENINetwork(c.primaryIP[eni], eniMetadata.MAC, eniMetadata.DeviceNumber, eniMetadata.SubnetIPv4CIDR)
 		if err != nil {
+			// Failed to set up the ENI
```

Comment:
How are you planning to test this? One way to test this would be add a sleep before `netLink.LinkSetUp` and delete the ENI and make sure eniConfig doesn't have this ENI.

Reply:
First of all we should add a unit test for it. 

For manual testing, deleting the ENI and check the logs and introspection endpoint.

Reply:
But you need to delete ENI in such a way that setupENI calls fails.

Reply:
Added debug code:
```
	if rand.Intn(2) == 0 {
		log.Debug("XXXX Failing randomly!")
		return errors.New("random linkByMac failure injected")
	}
```
Output:
```
{"level":"debug","ts":"2020-08-31T21:29:50.258Z","caller":"ipamd/ipamd.go:476","msg":"Reconcile existing ENI eni-01d689e8a83c6feba IP pool"}
{"level":"debug","ts":"2020-08-31T21:29:50.258Z","caller":"ipamd/ipamd.go:476","msg":"XXXXX Deleted the last IP from attachedENI: %+v"}
{"level":"warn","ts":"2020-08-31T21:29:50.258Z","caller":"ipamd/ipamd.go:968","msg":"Instance metadata does not match data store! ipPool: [192.168.159.31 192.168.134.205 192.168.152.224 192.168.134.35 192.168.129.132 192.168.134.155 192.168.149.75 192.168.128.237 192.168.150.15 192.168.147.91 192.168.156.197 192.168.131.24 192.168.133.50 192.168.152.119], metadata: [{\n  Primary: true,\n  PrivateIpAddress: \"192.168.131.160\"\n} {\n  Primary: false,\n  PrivateIpAddress: \"192.168.149.75\"\n} {\n  Primary: false,\n  PrivateIpAddress: \"192.168.134.205\"\n} {\n  Primary: false,\n  PrivateIpAddress: \"192.168.128.237\"\n} {\n  Primary: false,\n  PrivateIpAddress: \"192.168.150.15\"\n} {\n  Primary: false,\n  PrivateIpAddress: \"192.168.152.224\"\n} {\n  Primary: false,\n  PrivateIpAddress: \"192.168.134.35\"\n} {\n  Primary: false,\n  PrivateIpAddress: \"192.168.129.132\"\n} {\n  Primary: false,\n  PrivateIpAddress: \"192.168.156.197\"\n} {\n  Primary: false,\n  PrivateIpAddress: \"192.168.131.24\"\n} {\n  Primary: false,\n  PrivateIpAddress: \"192.168.134.155\"\n} {\n  Primary: false,\n  PrivateIpAddress: \"192.168.147.91\"\n} {\n  Primary: false,\n  PrivateIpAddress: \"192.168.159.31\"\n} {\n  Primary: false,\n  PrivateIpAddress: \"192.168.133.50\"\n}]"}
{"level":"debug","ts":"2020-08-31T21:29:50.258Z","caller":"ipamd/ipamd.go:968","msg":"We need to check the ENI status by calling the EC2 control plane."}
{"level":"debug","ts":"2020-08-31T21:29:50.352Z","caller":"ipamd/ipamd.go:1026","msg":"Reconcile and skip primary IP 192.168.131.160 on ENI eni-01d689e8a83c6feba"}
{"level":"debug","ts":"2020-08-31T21:29:50.352Z","caller":"ipamd/ipamd.go:476","msg":"Reconcile existing ENI eni-074c28aba8340f164 IP pool"}
{"level":"debug","ts":"2020-08-31T21:29:50.352Z","caller":"ipamd/ipamd.go:476","msg":"XXXXX Deleted the last IP from attachedENI: %+v"}
{"level":"warn","ts":"2020-08-31T21:29:50.352Z","caller":"ipamd/ipamd.go:968","msg":"Instance metadata does not match data store! ipPool: [192.168.145.78 192.168.145.183 192.168.133.18 192.168.155.72 192.168.143.232 192.168.132.174 192.168.145.239 192.168.133.98 192.168.157.218 192.168.143.209 192.168.153.245 192.168.139.151 192.168.150.195 192.168.150.150], metadata: [{\n  Primary: true,\n  PrivateIpAddress: \"192.168.153.162\"\n} {\n  Primary: false,\n  PrivateIpAddress: \"192.168.155.72\"\n} {\n  Primary: false,\n  PrivateIpAddress: \"192.168.143.232\"\n} {\n  Primary: false,\n  PrivateIpAddress: \"192.168.132.174\"\n} {\n  Primary: false,\n  PrivateIpAddress: \"192.168.145.78\"\n} {\n  Primary: false,\n  PrivateIpAddress: \"192.168.145.239\"\n} {\n  Primary: false,\n  PrivateIpAddress: \"192.168.133.98\"\n} {\n  Primary: false,\n  PrivateIpAddress: \"192.168.150.195\"\n} {\n  Primary: false,\n  PrivateIpAddress: \"192.168.157.218\"\n} {\n  Primary: false,\n  PrivateIpAddress: \"192.168.143.209\"\n} {\n  Primary: false,\n  PrivateIpAddress: \"192.168.133.18\"\n} {\n  Primary: false,\n  PrivateIpAddress: \"192.168.153.245\"\n} {\n  Primary: false,\n  PrivateIpAddress: \"192.168.150.150\"\n} {\n  Primary: false,\n  PrivateIpAddress: \"192.168.145.183\"\n}]"}
{"level":"debug","ts":"2020-08-31T21:29:50.352Z","caller":"ipamd/ipamd.go:968","msg":"We need to check the ENI status by calling the EC2 control plane."}
{"level":"debug","ts":"2020-08-31T21:29:50.453Z","caller":"ipamd/ipamd.go:1026","msg":"Reconcile and skip primary IP 192.168.153.162 on ENI eni-074c28aba8340f164"}
{"level":"debug","ts":"2020-08-31T21:29:50.453Z","caller":"ipamd/ipamd.go:476","msg":"Reconcile and add a new ENI {eni-0c57b7eecd642377d 06:ec:7b:a5:81:5f %!s(int=3) 192.168.128.0/19 [{\n  Primary: true,\n  PrivateIpAddress: \"192.168.159.161\"\n} {\n  Primary: false,\n  PrivateIpAddress: \"192.168.144.13\"\n} {\n  Primary: false,\n  PrivateIpAddress: \"192.168.154.110\"\n} {\n  Primary: false,\n  PrivateIpAddress: \"192.168.147.175\"\n} {\n  Primary: false,\n  PrivateIpAddress: \"192.168.138.96\"\n} {\n  Primary: false,\n  PrivateIpAddress: \"192.168.145.34\"\n} {\n  Primary: false,\n  PrivateIpAddress: \"192.168.146.227\"\n} {\n  Primary: false,\n  PrivateIpAddress: \"192.168.155.154\"\n} {\n  Primary: false,\n  PrivateIpAddress: \"192.168.137.157\"\n} {\n  Primary: false,\n  PrivateIpAddress: \"192.168.149.94\"\n} {\n  Primary: false,\n  PrivateIpAddress: \"192.168.142.62\"\n} {\n  Primary: false,\n  PrivateIpAddress: \"192.168.129.31\"\n} {\n  Primary: false,\n  PrivateIpAddress: \"192.168.149.50\"\n} {\n  Primary: false,\n  PrivateIpAddress: \"192.168.134.178\"\n} {\n  Primary: false,\n  PrivateIpAddress: \"192.168.140.51\"\n}]}"}
{"level":"debug","ts":"2020-08-31T21:29:50.453Z","caller":"ipamd/ipamd.go:742","msg":"DataStore Add an ENI eni-0c57b7eecd642377d"}
{"level":"info","ts":"2020-08-31T21:29:50.453Z","caller":"networkutils/network.go:704","msg":"Setting up network for an ENI with IP address 192.168.159.161, MAC address 06:ec:7b:a5:81:5f, CIDR 192.168.128.0/19 and route table 3"}
{"level":"debug","ts":"2020-08-31T21:29:50.453Z","caller":"networkutils/network.go:704","msg":"XXXX Failing randomly!"}
{"level":"info","ts":"2020-08-31T21:29:50.453Z","caller":"ipamd/ipamd.go:754","msg":"RemoveENIFromDataStore eni-0c57b7eecd642377d: IP address pool stats: free 0 addresses, total: 28, assigned: 17"}
{"level":"error","ts":"2020-08-31T21:29:50.453Z","caller":"ipamd/ipamd.go:476","msg":"IP pool reconcile: Failed to set up ENI eni-0c57b7eecd642377d network: failed to set up ENI eni-0c57b7eecd642377d network: random linkByMac failure injected"}
{"level":"debug","ts":"2020-08-31T21:29:50.453Z","caller":"ipamd/ipamd.go:476","msg":"Successfully Reconciled ENI/IP pool"}
{"level":"debug","ts":"2020-08-31T21:29:50.453Z","caller":"ipamd/ipamd.go:476","msg":"IP Address Pool stats: total: 28, assigned: 17"}
```